src/utils/udnsevent.cc

```diff
@@ -0,0 +1,215 @@
+
+#include <torrent/common.h>
+#include "udnsevent.h"
+#include "globals.h"
+#include "manager.h"
+#include "torrent/poll.h"
+
+namespace torrent {
+
+  int udnserror_to_gaierror(int udnserror) {
+    switch (udnserror) {
+      case DNS_E_TEMPFAIL:
+        return EAI_AGAIN;
+      case DNS_E_PROTOCOL:
+        // this isn't quite right
+        return EAI_FAIL;
+      case DNS_E_NXDOMAIN:
+        return EAI_NONAME;
+      case DNS_E_NODATA:
+        return EAI_ADDRFAMILY;
+      case DNS_E_NOMEM:
+        return EAI_MEMORY;
+      case DNS_E_BADQUERY:
+        return EAI_NONAME;
+      default:
+        return EAI_ADDRFAMILY;
+    }
+  }
+
+  /** Compatibility layers so udns can call std::function callbacks. */
+
+  void a4_callback_wrapper(struct ::dns_ctx *ctx, ::dns_rr_a4 *result, void *data) {
+    struct sockaddr_in sa;
+    const sockaddr *sa_addr = (sockaddr *) &sa;
+    UdnsQuery *query = (UdnsQuery *) data;
+    // udns will free the a4_query after this callback exits
+    query->a4_query = NULL;
+
+    if (result == NULL || result->dnsa4_nrr == 0) {
+      if (query->a6_query == NULL) {
+        // nothing more to do: call the callback with a failure status
+        (*(query->callback))(NULL, udnserror_to_gaierror(::dns_status(ctx)));
+        delete query;
+      }
+      // else: return and wait to see if we get an a6 response
+    } else {
+      sa.sin_family = AF_INET;
+      sa.sin_port = 0;
+      sa.sin_addr = result->dnsa4_addr[0];
+      if (query->a6_query != NULL) {
+        ::dns_cancel(ctx, query->a6_query);
+      }
+      (*query->callback)(sa_addr, 0);
+      delete query;
+    }
+  }
+
+  void a6_callback_wrapper(struct ::dns_ctx *ctx, ::dns_rr_a6 *result, void *data) {
+    struct sockaddr_in6 sa;
+    const sockaddr *sa_addr = (sockaddr *) &sa;
+    UdnsQuery *query = (UdnsQuery *) data;
+    // udns will free the a6_query after this callback exits
+    query->a6_query = NULL;
+
+    if (result == NULL || result->dnsa6_nrr == 0) {
+      if (query->a4_query == NULL) {
+        // nothing more to do: call the callback with a failure status
+        (*(query->callback))(NULL, udnserror_to_gaierror(::dns_status(ctx)));
+        delete query;
+      }
+      // else: return and wait to see if we get an a6 response
+    } else {
+      sa.sin6_family = AF_INET6;
+      sa.sin6_port = 0;
+      sa.sin6_addr = result->dnsa6_addr[0];
+      if (query->a4_query != NULL) {
+        ::dns_cancel(ctx, query->a4_query);
+      }
+      (*query->callback)(sa_addr, 0);
+      delete query;
+    }
+  }
+
+
+  UdnsEvent::UdnsEvent() {
+    // reinitialize the default context, no-op
+    // TODO don't do this here --- do it once in the manager, or in rtorrent
+    ::dns_init(NULL, 0);
+    // thread-safe context isolated to this object:
+    m_ctx = ::dns_new(NULL);
+    int fd = ::dns_init(m_ctx, 1);
+    if (fd > 0) {
+      m_fileDesc = fd;
+    } else {
+      throw internal_error("dns_init failed");
+    }
```

Comment:
Reorder this so that you only go into an if block in the exceptional case:

```
m_fileDesc = ::dns_init...;

if (fd == -1)
  throw ...
```

Also looking at the documentation it looks like if you call dns_init(m_ctx, 1) you don't need to call it above. Or if you do called it above, just use dns_open.